vm-setup/roles/v1aX_integration_test/tasks/node_reuse.yml

```diff
@@ -3,161 +3,180 @@
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-
+    k8s:
+      api_version: infrastructure.cluster.x-k8s.io/v1alpha4
+      kind: Metal3MachineTemplate
+      name: "{{ CLUSTER_NAME }}-controlplane"
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+      apply: yes
+      definition:
+        spec:
+          nodeReuse: True
+  
+  - name: Update maxSurge field to 0 and upgrade KubeadmControlPlane k8s version from "{{ KUBERNETES_VERSION }}" to "{{ UPGRADED_K8S_VERSION }}" to start testing KCP scale-in feature. 
+    k8s:
+      api_version: controlplane.cluster.x-k8s.io/v1alpha3
+      kind: KubeadmControlPlane
+      name: "{{ CLUSTER_NAME }}"
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+      apply: yes
+      definition:
+        spec:
+          rolloutStrategy:
+            rollingUpdate:
+              maxSurge: 0
+          version: "{{ UPGRADED_K8S_VERSION }}"
+  
   - pause:
       minutes: 1
-
-  - name: Check if any of the machines is in Provisioning state.
-    shell: |
-        kubectl get machines -n "{{ NAMESPACE }}" -o json | jq -r '[ .items[] | select (.status.phase == "Provisioning") | .metadata.name ] | length'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: provisioning_machine
-
+  
   - name: Check if only single machine is in Deleting state and no other new machine is in Provisioning state while upgrade triggered to test the KCP scale-in feature.
-    shell: |
-        kubectl get machines -n "{{ NAMESPACE }}" -o json | jq -r '[ .items[] | select (.status.phase == "Deleting" or .status.phase == "deleting") | .metadata.name ] | length'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: deleting_machines
+    k8s_info:
+      api_version: cluster.x-k8s.io/v1alpha3
+      kind: machines
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+      namespace: "{{ NAMESPACE }}"
+    register: provisioning_machine
     retries: 200
     delay: 2
-    until: deleting_machines.stdout == NUMBER_OF_BMH_MD
-    failed_when: provisioning_machine.stdout != "0"
-
-  - name: Wait until "{{ NUMBER_OF_BMH_MD }}" BMH is in deprovisioning state.
-    shell: |
-        kubectl get bmh -n "{{ NAMESPACE }}" -o json | jq -r '[ .items[]
-        | select (.status.provisioning.state == "deprovisioning")
-        | .metadata.name ] | length'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: bmh_in_deprovisioning
+    until: 
+      - provisioning_machine is succeeded
+      - provisioning_machine.resources | filter_phase("deleting") | length == (NUMBER_OF_BMH_MD | int)
+    failed_when: 
+      - provisioning_machine is succeeded
+      - provisioning_machine.resources | filter_phase("Provisioning") | length  != 0
+
```

Comment:
why we have `provisioning_machine is succeeded` in both the desired and failure conditions?

Reply:
I was trying to avoid api calls fails when resources does not exist ! Do not think this would cause an issue

Reply:
Hmm yes that looks weird... I think it should only be in `until`. Actually, now that I think about it, I'm not sure why we need `failed_when` at all? It was necessary when using `kubectl` since it can give an error code if there is some temporary problem with the connection, but now it should not be a problem since we use the module and have `provisioning_machine is succeeded` in `until`.

Reply:
IMHO `failed_when` ensures that `no other new machine is in Provisioning state while upgrade triggered` as mentioned in the task name we can use it in the until condition by inverting the condition to `provisioning_machine.resources | filter_phase("Provisioning") | length  == 0`. I did not get how it can be related to `kubectl` only